app/lib/user_annotation_service.rb

```diff
@@ -0,0 +1,102 @@
+      user_data_arrays_attributes.keys.each do |key|
+        data_names.push(user_data_arrays_attributes[key][:name].strip)
+      end
+
+      source_resolution = subsample_threshold.present? ? subsample_threshold.to_i : nil
+
+      # Create the annotation
+      user_annotation = UserAnnotation.new(user_id: user_id, study_id: study_id,
+                                            cluster_group_id: cluster_group_id,
+                                            values: data_names, name: name,
+                                            source_resolution: source_resolution)
+
+      # override cluster setter to use the current selected cluster, needed for reloading
+      cluster = user_annotation.cluster_group
+
+      # Save the user annotation, and handle any exceptions
+      if user_annotation.save
+
+        # Consider refactoring user_annotation_test.rb and
+        # models/user_annotation.rb to remove `subsample_annotation`.  Jon
+        # and Eric investigated on 2021-01-22 and determined that there are no
+        # known use cases where `subsample_annotation` != `loaded_annotation`.
+        subsample_annotation = loaded_annotation
+
+        # Method call to create the user data arrays for this annotation
+        user_annotation.initialize_user_data_arrays(user_data_arrays_attributes, subsample_annotation, subsample_threshold, loaded_annotation)
+
+        # Reset the annotations in the dropdowns to include this new annotation
+        cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+
+        message = "User Annotation: '#{user_annotation.name}' successfully saved. You may now view this annotation via the annotations dropdown."
+        status = 200 # Success
+      else
+        # If there was an error saving, alert the user something broke
+        cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+        message = 'The following errors prevented the annotation from being saved: ' + user_annotation.errors.full_messages.join(',')
+        Rails.logger.error "Creating user annotation of params: #{user_annotation_params}, unable to save user annotation with errors #{user_annotation.errors.full_messages.join(', ')}"
+        status = 400  # Bad request
+      end
+      [message, cluster_annotations, status]
+
+    # Handle other errors in saving user annotation
+    rescue Mongoid::Errors::InvalidValue => e
+      error_context = ErrorTracker.format_extra_context(study, {params: log_params})
+      ErrorTracker.report_exception(e, current_user, error_context)
+      # If an invalid value was somehow passed through the form, and couldn't save the annotation
+      cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+      message = 'The following errors prevented the annotation from being saved: ' + 'Invalid data type submitted. (' + e.problem + '. ' + e.resolution + ')'
```

Comment:
I like that you're catching a relatively low-level message here and reformatting it to be more sensible.  Instead of a complex return type for this function though, I think you should just raise a new exception with your more informative message

Reply:
Done -- latest code refines this aspect of the old `site_controller.rb` handling.